// @anyone-protocol/anyone-client/out/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Flag = exports.EventType = void 0;
var EventType;
(function (EventType) {
    EventType["ADDRMAP"] = "ADDRMAP";
    EventType["BUILDTIMEOUT_SET"] = "BUILDTIMEOUT_SET";
    EventType["BW"] = "BW";
    EventType["CELL_STATS"] = "CELL_STATS";
    EventType["CIRC"] = "CIRC";
    EventType["CIRC_BW"] = "CIRC_BW";
    EventType["CIRC_MINOR"] = "CIRC_MINOR";
    EventType["CONF_CHANGED"] = "CONF_CHANGED";
    EventType["CONN_BW"] = "CONN_BW";
    EventType["CLIENTS_SEEN"] = "CLIENTS_SEEN";
    EventType["DEBUG"] = "DEBUG";
    EventType["DESCCHANGED"] = "DESCCHANGED";
    EventType["ERR"] = "ERR";
    EventType["GUARD"] = "GUARD";
    EventType["HS_DESC"] = "HS_DESC";
    EventType["HS_DESC_CONTENT"] = "HS_DESC_CONTENT";
    EventType["INFO"] = "INFO";
    EventType["NETWORK_LIVENESS"] = "NETWORK_LIVENESS";
    EventType["NEWCONSENSUS"] = "NEWCONSENSUS";
    EventType["NEWDESC"] = "NEWDESC";
    EventType["NOTICE"] = "NOTICE";
    EventType["NS"] = "NS";
    EventType["ORCONN"] = "ORCONN";
    EventType["SIGNAL"] = "SIGNAL";
    EventType["STATUS_CLIENT"] = "STATUS_CLIENT";
    EventType["STATUS_GENERAL"] = "STATUS_GENERAL";
    EventType["STATUS_SERVER"] = "STATUS_SERVER";
    EventType["STREAM"] = "STREAM";
    EventType["STREAM_BW"] = "STREAM_BW";
    EventType["TRANSPORT_LAUNCHED"] = "TRANSPORT_LAUNCHED";
    EventType["WARN"] = "WARN";
    EventType["UNKNOWN"] = "UNKNOWN";
})(EventType || (exports.EventType = EventType = {}));
var Flag;
(function (Flag) {
    Flag["Authority"] = "Authority";
    Flag["BadExit"] = "BadExit";
    Flag["BadDirectory"] = "BadDirectory";
    Flag["Exit"] = "Exit";
    Flag["Fast"] = "Fast";
    Flag["Guard"] = "Guard";
    Flag["HSDir"] = "HSDir";
    Flag["Named"] = "Named";
    Flag["NoEdConsensus"] = "NoEdConsensus";
    Flag["Running"] = "Running";
    Flag["Stable"] = "Stable";
    Flag["StaleDesc"] = "StaleDesc";
    Flag["Unnamed"] = "Unnamed";
    Flag["Valid"] = "Valid";
    Flag["V2Dir"] = "V2Dir";
    Flag["V3Dir"] = "V3Dir";
})(Flag || (exports.Flag = Flag = {}));
